### PR TITLE
libffi => 3.4.2 & python3 => 3.10.1, and all downstream packages requiring updates

### PR DESCRIPTION
**Pull request by saltedcoffii**
Upgrade of libffi to 3.4.2 & python to 3.10
Before we merge this pull request, it is imperative that **ALL** packages depending on python3 for build or runtime rebuild and preferably upgraded to their latest version. I can help with this but would love backup, either from a human, or if anyone knows of an update checker script, that would be really helpful as well.

All packages requiring libffi also need updating.

This will tell you what needs updating on a `x86_64` system for the libffi update:
```for i in /usr/local/lib64/*.so ; do ldd $i 2>/dev/null | grep libffi.so.7 &>/dev/null && crew whatprovides $i | head -n 1 | sed 's/://' | awk '{print $1'} ; done | sort -u```

**All packages in the list that command gives need to be updated in this PR.**

That list may include at a minimum the following, but this list may shrink as common dependencies get updated.

- [x] cairo
- [x] cups
- [x] ffmpeg
- [x] gjs
- [x] glib
- [x] gnutls
- [x] gobject_introspection
- [x] llvm
- [x] p11kit
- [x] rtmpdump
- [x] wayland

### Run the following to get this pull request's changes locally for testing.
```
CREW_TESTING_REPO=https://github.com/saltedcoffii/chromebrew.git CREW_TESTING_BRANCH=python3.10 CREW_TESTING=1 crew update
```

### CODE REVIEW

**Comment by saltedcoffii** [comment]
@satmandu how are the binaries coming?

**Comment by uberhacker** [comment]
Don't you need to `crew reinstall libffi -s` as the first step?

**Comment by uberhacker** [comment]
Encountered this error while building python3 on i686:
```
Stripping libraries...
llvm-strip: error while loading shared libraries: libffi.so.7: cannot open shared object file: No such file or directory
Stripping binaries...
llvm-strip: error while loading shared libraries: libffi.so.7: cannot open shared object file: No such file or directory
Archiving...
python3 is built!
```

**Comment by uberhacker** [comment]
Attached is the bash shell script I used to build the binaries.

**Comment by uberhacker** [comment]
@satmandu:  The python 3.10 executable runs fine on i686 but clang and clang++ still segfault with the llvm update.  Whatever you did with the iptables update, keep doing it!

**Comment by saltedcoffii** [comment]
@satmandu @uberhacker what's the status on this pull request?

**Comment by uberhacker** [comment]
Unfortunately, the binaries built with the latest changes to llvm_stage1 still segfault on i686.  See below:
```
chronos@localhost ~/chromebrew/packages (python3.10 *=)$ crew reinstall llvm_stage1
llvm_stage1: THIS IS NOT THE LLVM PACKAGE. This is only a Limited Stage 1 compile of LLVM.
Performing pre-flight checks...
Precompiled binary available, downloading...
##################################################################################################################################### 100.0%
Llvm_stage1 archive downloaded.
Unpacking 'tpxz' archive using 'tar', this may take a while...
Removing since upgrade or reinstall...
Llvm_stage1 removed!
Performing pre-install...
Performing install...
Performing post-install for llvm_stage1...

To compile programs, use 'clang' or 'clang++'.

Llvm_stage1 installed!
chronos@localhost ~/chromebrew/packages (python3.10 *=)$ crew search llvm_stage1 -v
llvm_stage1: THIS IS NOT THE LLVM PACKAGE. This is only a Limited Stage 1 compile of LLVM.
http://llvm.org/
Version: 13.0.0
License: Apache-2.0-with-LLVM-exceptions, UoI-NCSA, BSD, public-domain and rc
chronos@localhost ~/chromebrew/packages (python3.10 *=)$ clang -v
Segmentation fault (core dumped)
chronos@localhost ~/chromebrew/packages (python3.10 *=)$ clang++ -v
Segmentation fault (core dumped)
```
I even reinstalled libffi to make sure everything was updated.

**Comment by satmandu** [comment]
@saltedcoffii I pushed the wireshark PR into here because it uses libffi, and I don't want to have to build it again once this PR is done. (I'm rebuilding those packages with this libffi now.)

**Comment by satmandu** [comment]
(Also there are vulkan & mesa updates here not in master... please note when merging master.)

**Comment by satmandu** [comment]
Still need to test the new `i686` builds here, but I think all compiles & rebuilds are done for this PR. Please test!

**Comment by satmandu** [comment]
This is good to go.

**Comment by satmandu** [comment]
@saltedcoffii @uberhacker @supechicken  Please test this? Does it break anything?

**Comment by saltedcoffii** [comment]
I've been running on this branch for about a month, intermittently merging with master and testing and I'm happy to say that it's been working for me. More specifically, `ffmpeg`, `firefox` and `python3` work. are there any specific packages you'd like me to test?

**Comment by satmandu** [comment]
@saltedcoffii I updated `ffmpeg` rebuilt `xwayland` earlier today. If those work for you...

**Comment by saltedcoffii** [comment]
`xwayland` and `ffmpeg` both work for me

**Comment by satmandu** [comment]
@saltedcoffii py3_numpy does not build...

**Comment by satmandu** [comment]
```
Traceback (most recent call last):
  File "/usr/local/tmp/crew/py3_numpy.20211208060856.dir/setup.py", line 57, in <module>
    MAJOR, MINOR, MICRO = re.match(r'(\d+)\.(\d+)\.(\d+)', FULLVERSION).groups()
AttributeError: 'NoneType' object has no attribute 'groups'
py3_numpy failed to build: `python3 setup.py build --executable=/usr/local/bin/python3` exited with 1
```

**Comment by saltedcoffii** [comment]
> ```
> Traceback (most recent call last):
>   File "/usr/local/tmp/crew/py3_numpy.20211208060856.dir/setup.py", line 57, in <module>
>     MAJOR, MINOR, MICRO = re.match(r'(\d+)\.(\d+)\.(\d+)', FULLVERSION).groups()
> AttributeError: 'NoneType' object has no attribute 'groups'
> �py3_numpy failed to build: `python3 setup.py build --executable=/usr/local/bin/python3` exited with 1
> ```

Looks like this is a known issue. Numpy just doesn't support python3.10 yet. What do you suggest @satmandu?

**Comment by satmandu** [comment]
Looks like we need `crew` to support git tags...  I have a commit which should handle that.....

Currently building with that and 
```
require 'package'

class Py3_numpy < Package
  description 'NumPy is the fundamental package for array computing with Python.'
  homepage 'https://numpy.org/'
  @_ver = '1.21.4'
  version @_ver
compatibility 'all'
  license 'BSD'
  source_url 'https://github.com/numpy/numpy.git'
  git_hashtag "v#{@_ver}"

  depends_on 'lapack'
  depends_on 'py3_cython' => :build
  depends_on 'py3_setuptools' => :build

  def self.build
    system "python3 setup.py build #{PY3_SETUP_BUILD_OPTIONS}"
  end

  def self.install
    system "python3 setup.py install #{PY_SETUP_INSTALL_OPTIONS}"
  end
end
```

File: packages/asciinema.rb [hunk 1]
```diff
@@ -3,24 +3,12 @@
...
-     armv7l: 'https://gitlab.com/api/v4/projects/26210301/packages/generic/asciinema/2.0.2-1_armv7l/asciinema-2.0.2-1-chromeos-armv7l.tar.xz',
-       i686: 'https://gitlab.com/api/v4/projects/26210301/packages/generic/asciinema/2.0.2-2_i686/asciinema-2.0.2-2-chromeos-i686.tpxz',
-     x86_64: 'https://gitlab.com/api/v4/projects/26210301/packages/generic/asciinema/2.0.2-2_x86_64/asciinema-2.0.2-2-chromeos-x86_64.tpxz'
-  })
-  binary_sha256({
-    aarch64: '5fddd233f9f093935c23574dc1ce6247103352e845408375fa537fd2f56675ec',
-     armv7l: '5fddd233f9f093935c23574dc1ce6247103352e845408375fa537fd2f56675ec',
-       i686: '9086a6bce66d3ad0542d8d61f6904bd24078998d2a0867ced9b65eabb15b889b',
-     x86_64: 'a94b73de95817df65ede7ba0e58c71547cf7eb6a326b9707dd94a29a3fd73aab'
-  })
```
Comment (uberhacker) [inline packages/asciinema.rb:23]: No binaries?  Building from git is not normal user-friendly.

File: packages/libxcb.rb [hunk 2]
```diff
@@ -3,23 +3,19 @@
...
   source_url 'https://xcb.freedesktop.org/dist/libxcb-1.14.tar.xz'
   source_sha256 'a55ed6db98d43469801262d81dc2572ed124edc3db31059d4e9916eb9f844c34'
 
   binary_url({
-    aarch64: 'https://gitlab.com/api/v4/projects/26210301/packages/generic/libxcb/1.14-2_armv7l/libxcb-1.14-2-chromeos-armv7l.tar.xz',
-     armv7l: 'https://gitlab.com/api/v4/projects/26210301/packages/generic/libxcb/1.14-2_armv7l/libxcb-1.14-2-chromeos-armv7l.tar.xz',
-       i686: 'https://gitlab.com/api/v4/projects/26210301/packages/generic/libxcb/1.14-2_i686/libxcb-1.14-2-chromeos-i686.tar.xz',
-     x86_64: 'https://gitlab.com/api/v4/projects/26210301/packages/generic/libxcb/1.14-2_x86_64/libxcb-1.14-2-chromeos-x86_64.tar.xz'
+    i686: 'https://gitlab.com/api/v4/projects/26210301/packages/generic/libxcb/1.14-3_i686/libxcb-1.14-3-chromeos-i686.tpxz',
+  x86_64: 'https://gitlab.com/api/v4/projects/26210301/packages/generic/libxcb/1.14-3_x86_64/libxcb-1.14-3-chromeos-x86_64.tpxz'
```
Comment (uberhacker) [inline packages/libxcb.rb:14]: No arm binaries?

File: packages/mutter.rb [hunk 3]
```diff
@@ -3,21 +3,21 @@
...
   source_url 'https://gitlab.gnome.org/GNOME/mutter.git'
   git_hashtag version
 
   binary_url({
-    aarch64: 'https://gitlab.com/api/v4/projects/26210301/packages/generic/mutter/40.3_armv7l/mutter-40.3-chromeos-armv7l.tpxz',
-     armv7l: 'https://gitlab.com/api/v4/projects/26210301/packages/generic/mutter/40.3_armv7l/mutter-40.3-chromeos-armv7l.tpxz',
-     x86_64: 'https://gitlab.com/api/v4/projects/26210301/packages/generic/mutter/40.3_x86_64/mutter-40.3-chromeos-x86_64.tpxz'
+    aarch64: 'https://gitlab.com/api/v4/projects/26210301/packages/generic/mutter/41.0_armv7l/mutter-41.0-chromeos-armv7l.tpxz',
+     armv7l: 'https://gitlab.com/api/v4/projects/26210301/packages/generic/mutter/41.0_armv7l/mutter-41.0-chromeos-armv7l.tpxz',
+     x86_64: 'https://gitlab.com/api/v4/projects/26210301/packages/generic/mutter/41.0_x86_64/mutter-41.0-chromeos-x86_64.tpxz'
```
Comment (uberhacker) [inline packages/mutter.rb:15]: No i686 binary?

File: packages/oci_cli.rb [hunk 4]
```diff
@@ -13,14 +13,14 @@ class Oci_cli < Package
   binary_url({
     aarch64: 'https://gitlab.com/api/v4/projects/26210301/packages/generic/oci_cli/2.24.3_armv7l/oci_cli-2.24.3-chromeos-armv7l.tpxz',
      armv7l: 'https://gitlab.com/api/v4/projects/26210301/packages/generic/oci_cli/2.24.3_armv7l/oci_cli-2.24.3-chromeos-armv7l.tpxz',
-       i686: 'https://gitlab.com/api/v4/projects/26210301/packages/generic/oci_cli/2.24.3_i686/oci_cli-2.24.3-chromeos-i686.tpxz',
-     x86_64: 'https://gitlab.com/api/v4/projects/26210301/packages/generic/oci_cli/2.24.3_x86_64/oci_cli-2.24.3-chromeos-x86_64.tpxz'
+    i686: 'https://gitlab.com/api/v4/projects/26210301/packages/generic/oci_cli/3.1.2_i686/oci_cli-3.1.2-chromeos-i686.tpxz',
+  x86_64: 'https://gitlab.com/api/v4/projects/26210301/packages/generic/oci_cli/3.1.2_x86_64/oci_cli-3.1.2-chromeos-x86_64.tpxz'
```
Comment (uberhacker) [inline packages/oci_cli.rb:17]: The arm binary version doesn't match the package version?

File: packages/mutter.rb (hunk 3 above)
Comment (satmandu) [reply packages/mutter.rb:15]: i686 is no longer supported.

File: packages/libxcb.rb (hunk 2 above)
Comment (satmandu) [reply packages/libxcb.rb]: fixed

File: packages/asciinema.rb (hunk 1 above)
Comment (satmandu) [reply packages/asciinema.rb:23]: This took less than 1 min to build...

File: packages/oci_cli.rb (hunk 4 above)
Comment (satmandu) [reply packages/oci_cli.rb:17]: Yup. Rebuilding this one on arm...

Comment (satmandu) [reply packages/oci_cli.rb]: done

File: packages/asciinema.rb (hunk 1 above)
Comment (satmandu) [reply packages/asciinema.rb:23]: binaries added

File: packages/mutter.rb (hunk 3 above)
Comment (satmandu) [reply packages/mutter.rb:15]: (no longer supported upstream)

File: packages/py3_py.rb [hunk 5]
```diff
@@ -17,10 +17,10 @@ class Py3_py < Package
...
   })
   binary_sha256({
-    aarch64: '87216dfa6e6dea6bf5d2a2b706a054db9affbb78413de08b14df23711eeeb053',
-     armv7l: '87216dfa6e6dea6bf5d2a2b706a054db9affbb78413de08b14df23711eeeb053',
-       i686: 'f9b7e26edf6d2a5b605cf16c0ef76ad590b07472f7d9741b635286794de3ff0b',
-     x86_64: 'daf69c756cec7668d0a3aea2d2dd54e7634a5a0650a6452b8a3e004f001571df'
+    aarch64: '99bc10e10ff6dae52fcc1865c397a812b476cc8f2593576180b76e41aeb857db',
+     armv7l: '99bc10e10ff6dae52fcc1865c397a812b476cc8f2593576180b76e41aeb857db',
+       i686: 'd2e51388ffca680ac1d5b5dd4d4ed3889cdf386d8272b867852be2c48427e2a9',
+     x86_64: '335654f047aad9a13d9b82650c7fb55d892e42e56f6987a3963d85bf8b72d619'
```
Comment (uberhacker) [inline packages/py3_py.rb:23]: Bump version?

File: packages/py3_sphinxcontrib_devhelp.rb [hunk 6]
```diff
@@ -17,10 +17,10 @@ class Py3_sphinxcontrib_devhelp < Package
...
   })
   binary_sha256({
-    aarch64: 'dac0c64018d2773024219781f16dbc2d6f66313159eb805fea7426a942a03f36',
-     armv7l: 'dac0c64018d2773024219781f16dbc2d6f66313159eb805fea7426a942a03f36',
-       i686: '6c84e7d96a26722635d98ea7200c478b169152ace4d59a286484b947ef018936',
-     x86_64: '419c47c6c8b0c1e11ae84f4d9bcc6fe4e7afb8b00da07de21198fafe703a47f5'
+    aarch64: '3196686ec01222742cb48ee585aa5060203f7238a0636f691dd70d19c4ea68c1',
+     armv7l: '3196686ec01222742cb48ee585aa5060203f7238a0636f691dd70d19c4ea68c1',
+       i686: 'dbbea0432ac6ce25922765f23d2e4ee98dfc3a89f7175c5353df32ff19dcd3be',
+     x86_64: 'a325888dce188a29eb4a91c01819c16261e6c1f9cd88478ecf0b8e316a180462'
```
Comment (uberhacker) [inline packages/py3_sphinxcontrib_devhelp.rb:23]: Bump version?

**Comment by satmandu** [comment]
Building python 3.10.1 overnight...

File: packages/py3_py.rb (hunk 5 above)
Comment (satmandu) [reply packages/py3_py.rb]: Done.

File: packages/py3_sphinxcontrib_devhelp.rb (hunk 6 above)
Comment (satmandu) [reply packages/py3_sphinxcontrib_devhelp.rb]: Done.

**Comment by satmandu** [comment]
Python 3 updated to 3.10.1.

**Comment by saltedcoffii** [comment]
It feels so nice to finally have this done, thank you everybody, especially @satmandu for their work on this!!

**Pull request merged**